corporate/tests/test_stripe.py

```diff
@@ -6547,6 +6547,88 @@ def test_invoice_plans_as_needed(self, *mocks: Mock) -> None:
         invoice_plans_as_needed(add_months(self.next_month, 1))
         self.assert_length(outbox, messages_count + 1)
 
+    @responses.activate
+    @mock_stripe()
+    def test_invoice_scheduled_upgrade_realm_legacy_plan(self, *mocks: Mock) -> None:
+        remote_server = RemoteZulipServer.objects.get(hostname="demo.example.com")
+        server_billing_session = RemoteServerBillingSession(remote_server=remote_server)
+
+        # Migrate server to legacy plan.
+        with time_machine.travel(self.now, tick=False):
+            start_date = timezone_now()
+            end_date = add_months(start_date, months=3)
+            server_billing_session.migrate_customer_to_legacy_plan(start_date, end_date)
+
+        # Upload data.
+        self.add_mock_response()
+        with time_machine.travel(self.now, tick=False):
+            send_server_data_to_push_bouncer(consider_usage_statistics=False)
+
+        self.login("hamlet")
+        hamlet = self.example_user("hamlet")
+
+        # Login. Performs customer migration from server to realms.
+        self.execute_remote_billing_authentication_flow(hamlet)
+        remote_server.refresh_from_db()
+
+        # Schedule upgrade to business plan
+        with time_machine.travel(self.now, tick=False):
+            stripe_customer = self.add_card_and_upgrade(
+                remote_server_plan_start_date="billing_cycle_end_date", talk_to_stripe=False
+            )
+
+        zulip_realm_customer = Customer.objects.get(stripe_customer_id=stripe_customer.id)
+        assert zulip_realm_customer is not None
+        zulip_realm_plan = get_current_plan_by_customer(zulip_realm_customer)
+        assert zulip_realm_plan is not None
+        self.assertEqual(zulip_realm_plan.tier, CustomerPlan.TIER_SELF_HOSTED_LEGACY)
+        self.assertEqual(zulip_realm_plan.status, CustomerPlan.SWITCH_PLAN_TIER_AT_PLAN_END)
+        self.assertEqual(zulip_realm_plan.end_date, end_date)
+
+        new_plan = self.billing_session.get_next_plan(zulip_realm_plan)
+        assert new_plan is not None
+        self.assertEqual(new_plan.tier, CustomerPlan.TIER_SELF_HOSTED_BUSINESS)
+        self.assertEqual(new_plan.status, CustomerPlan.NEVER_STARTED)
+        self.assertEqual(new_plan.billing_cycle_anchor, end_date)
```

Comment:
Might as well assert the `invoicing_status` too.

Reply:
Done